File: clients/cpp/RemoteAPIClient.cpp
#include "RemoteAPIClient.h"
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
#include <boost/format.hpp>
#include <boost/optional.hpp>
#include <random>
#include <sstream>

namespace uuid {
    static std::random_device              rd;
    static std::mt19937                    gen(rd());
    static std::uniform_int_distribution<> dis(0, 15);
    static std::uniform_int_distribution<> dis2(8, 11);

    std::string generate_uuid_v4() {
        std::stringstream ss;
        int i;
        ss << std::hex;
        for (i = 0; i < 8; i++) {
            ss << dis(gen);
        }
        ss << "-";
        for (i = 0; i < 4; i++) {
            ss << dis(gen);
        }
        ss << "-4";
        for (i = 0; i < 3; i++) {
            ss << dis(gen);
        }
        ss << "-";
        ss << dis2(gen);
        for (i = 0; i < 3; i++) {
            ss << dis(gen);
        }
        ss << "-";
        for (i = 0; i < 12; i++) {
            ss << dis(gen);
        };
        return ss.str();
    }
}

json bin(const char *s, int size)
{
    if(size == -1) size = int(strlen(s));
    return bin(reinterpret_cast<const uint8_t *>(s), size);
}

json bin(const uint8_t *b, int size)
{
    return bin(std::vector<uint8_t>{b, b + size});
}

json bin(const std::string &s)
{
    return bin(s.data(), int(s.length()));
}

json bin(const std::vector<uint8_t> &v)
{
    return json{byte_string_arg, v};
}

RemoteAPIClient::RemoteAPIClient(const std::string host, int rpcPort, int cntPort, int verbose_)
    : rpcSocket(ctx, zmq::socket_type::req),
      verbose(verbose_)
{
    if(verbose == -1)
    {
        if(const char* verboseStr = std::getenv("VERBOSE"))
            verbose = std::atoi(verboseStr);
        else
            verbose = 0;
    }

    uuid = uuid::generate_uuid_v4();
    VERSION = 2;

    auto rpcAddr = (boost::format("tcp://%s:%d") % host % rpcPort).str();
    rpcSocket.connect(rpcAddr);
}

RemoteAPIClient::~RemoteAPIClient()
{
    json req;
    req["func"] = "_*end*_";
    req["args"] = json::array();
    send(req);
    recv();
}

json RemoteAPIClient::call(const std::string &func, std::initializer_list<json> args)
{
    return call(func, json::make_array(args));
}

json RemoteAPIClient::call(const std::string &_func, const json &_args)
{ // call function with specified arguments. Is reentrant
    std::string func(_func);
    json args = _args;
    json req;
    req["func"] = func;
    req["args"] = args;
    send(req);
    json resp = recv();

    while (resp.contains("func"))
    { // We have a callback or a wait:
        if (resp["func"].as<std::string>().compare("_*wait*_")==0)
        {
            func = "_*executed*_";
            args = json::array();
            json req2;
            req2["func"] = func;
            req2["args"] = args;
            send(req2);
        }
        else if (resp["func"].as<std::string>().compare("_*repeat*_")==0)
        {
            json req2;
            req2["func"] = func;
            req2["args"] = args;
            send(req2);
        }
        else
        { // call a callback
            auto funcToRun = _getFunctionPointerByName(resp["func"].as<std::string>());
            json rep;
            func = "_*executed*_";
            if (funcToRun)
            {
                auto r = funcToRun(resp["args"]);
                if (!r.is_array())
                {
                    auto arr = json::array();
                    arr.push_back(r);
                    args = arr;
                }
                else
                    args = r;
            }
            else
                args = json::array();
            rep["func"] = func;
            rep["args"] = args;
            send(rep);
        }
        resp = recv();
    }

    if (resp.contains("err"))
        throw std::runtime_error(resp["err"].as<std::string>().c_str());
    const auto &ret = resp["ret"];
    return ret;
}

json RemoteAPIClient::getObject(const std::string &name)
{
    return call("zmqRemoteApi.info", json(json_array_arg, {name}));
}

void RemoteAPIClient::require(const std::string &name)
{
    call("zmqRemoteApi.require", json(json_array_arg, {name}));
}

void RemoteAPIClient::setVerbose(int level)
{
    verbose = level;
}

void RemoteAPIClient::setStepping(bool enable)
{ // for backw. comp., now via sim.setStepping
    call("sim.setStepping", {enable});
}

void RemoteAPIClient::step(bool wait)
{ // for backw. comp., now via sim.step
    call("sim.step", {wait});
}

void RemoteAPIClient::send(json &j)
{
    if(verbose > 0)
        std::cout << "Sending: " << pretty_print(j) << std::endl;

    j["uuid"] = uuid;
    j["ver"] = VERSION;
    j["lang"] = "c++";
    if (j.contains("args"))
    {
        auto a = j["args"];
        j["argsL"] = a.size();
    }


    std::vector<uint8_t> data;
    cbor::encode_cbor(j, data);

    if(verbose > 1)
    {
        std::cout << "Sending (raw):";
        for(size_t i = 0; i < data.size(); i++)
            std::cout << " " << std::hex << std::setw(2) << std::setfill('0') << int(data[i]);
        std::cout << std::endl;
    }

    zmq::message_t msg(data.data(), data.size());
    rpcSocket.send(msg, zmq::send_flags::dontwait);
}

json RemoteAPIClient::recv()
{
    zmq::message_t msg;
    rpcSocket.recv(msg);

    auto data = reinterpret_cast<const uint8_t*>(msg.data());

    if(verbose > 1)
    {
        std::cout << "Received (raw):";
        for(size_t i = 0; i < msg.size(); i++)
            std::cout << " " << std::hex << std::setw(2) << std::setfill('0') << int(data[i]);
        std::cout << std::endl;
    }

    json j = cbor::decode_cbor<json>(data, data + msg.size());

    if(verbose > 0)
        std::cout << "Received: " << pretty_print(j) << std::endl;

    return j;
}

void RemoteAPIClient::registerCallback(const std::string &funcName, CallbackType callback)
{
    callbacks[funcName] = callback;
}

RemoteAPIClient::CallbackType RemoteAPIClient::_getFunctionPointerByName(const std::string &funcName)
{
    auto it = callbacks.find(funcName);
    if(it != callbacks.end())
        return it->second;
    return nullptr;
}

#ifdef SIM_REMOTEAPICLIENT_OBJECTS
#include "RemoteAPIObjects.cpp"
#endif // SIM_REMOTEAPICLIENT_OBJECTS

File: clients/java/src/main/java/com/coppeliarobotics/remoteapi/zmq/RemoteAPIClient.java
package com.coppeliarobotics.remoteapi.zmq;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.UUID;
import java.util.function.Function;

import java.math.BigInteger;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.IOException;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import co.nstant.in.cbor.*;
import co.nstant.in.cbor.model.DataItem;

import org.zeromq.*;

public class RemoteAPIClient
{
    public RemoteAPIClient()
    {
        this("localhost", -1);
    }

    public RemoteAPIClient(String host, int rpcPort)
    {
        this(host, rpcPort, -1);
    }

    public RemoteAPIClient(String host, int rpcPort, int cntPort)
    {
        this(host, rpcPort, cntPort, -1);
    }

    public RemoteAPIClient(String host, int rpcPort, int cntPort, int verbose)
    {
        if(rpcPort == -1) rpcPort = 23000;

        this.verbose = verbose;
        if(this.verbose == -1)
        {
            String verboseStr = System.getenv("VERBOSE");
            if(verboseStr != null)
                this.verbose = Integer.parseInt(verboseStr);
            else
                this.verbose = 0;
        }

        this.context = new ZContext(1);

        this.rpcSocket = context.createSocket(SocketType.REQ);
        this.rpcSocket.connect(String.format("tcp://%s:%d", host, rpcPort));

        this.uuid = UUID.randomUUID().toString();
    }

    public void close() throws CborException
    {
        DataItem k_func = convertArg("func"),
                 k_args = convertArg("args"),
                 k_err = convertArg("err"),
                 k_uuid = convertArg("uuid"),
                 k_ver = convertArg("ver"),
                 k_lang = convertArg("lang"),
                 k_ret = convertArg("ret"),
                 k_argsl = convertArg("argsL");

        co.nstant.in.cbor.model.Map req = new co.nstant.in.cbor.model.Map();
        req.put(k_func, convertArg("_*end*_"));
        req.put(k_args, new co.nstant.in.cbor.model.Array());
        req.put(k_uuid, convertArg(this.uuid));
        req.put(k_ver, convertArg(this.VERSION));
        this.send(req);
        this.recv();
        this.rpcSocket.close();
        this.context.close();
    }

    protected void send(DataItem req) throws CborException
    {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        new CborEncoder(baos).encode(req);
        byte[] data = baos.toByteArray();

        if(this.verbose >= 3)
            System.out.println("Sending (raw): " + dump(data, verbose >= 4 ? -1 : 16));

        this.rpcSocket.send(data);
    }

    protected DataItem recv() throws CborException
    {
        byte[] data = this.rpcSocket.recv(0);

        if(this.verbose >= 3)
            System.out.println("Received (raw): " + dump(data, verbose >= 4 ? -1 : 16));

        ByteArrayInputStream bais = new ByteArrayInputStream(data);
        List<DataItem> dataItems = new CborDecoder(bais).decode();
        return dataItems.get(0);
    }

    public String dump(byte[] data, int limit)
    {
        if(limit < 0)
            limit = Integer.MAX_VALUE;
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < Math.min(limit, data.length); i++)
            sb.append(String.format("%02X ", data[i]));
        if(data.length > limit)
            sb.append(String.format("(%d more)", data.length - limit));
        return sb.toString();
    }

    public List<DataItem> callCbor(String func, List<DataItem> args) throws CborException
    {
        DataItem k_func = convertArg("func"),
                 k_args = convertArg("args"),
                 k_err = convertArg("err"),
                 k_uuid = convertArg("uuid"),
                 k_ver = convertArg("ver"),
                 k_lang = convertArg("lang"),
                 k_ret = convertArg("ret"),
                 k_argsl = convertArg("argsL");

        co.nstant.in.cbor.model.Map req = new co.nstant.in.cbor.model.Map();
        req.put(k_func, convertArg(func));
        co.nstant.in.cbor.model.Array v_args = new co.nstant.in.cbor.model.Array();
        for(int i = 0; i < args.size(); i++)
            v_args.add(args.get(i));
        req.put(k_args, v_args);
        req.put(k_uuid, convertArg(this.uuid));
        req.put(k_ver, convertArg(this.VERSION));
        req.put(k_argsl, convertArg(args.size()));
        req.put(k_lang, convertArg("java"));
        if(this.verbose >= 1)
            System.out.println("Sending: " + req.toString());
        this.send(req);

        co.nstant.in.cbor.model.Map rep = (co.nstant.in.cbor.model.Map)this.recv();
        if(this.verbose >= 1)
            System.out.println("Received: " + rep.toString());

        while(rep.getKeys().contains(k_func))
        {
            // Check if the function name is "_*wait*_", in which case we just wait
            List<DataItem> callbackResults = new ArrayList<DataItem>();
            co.nstant.in.cbor.model.UnicodeString respFunc = (co.nstant.in.cbor.model.UnicodeString)rep.get(k_func);
            // _*repeat*_ not yet implemented!
            if(respFunc != null && !"_*wait*_".equals(respFunc.getString()))
            {
                List<DataItem> callbackArgs = ((co.nstant.in.cbor.model.Array)rep.get(k_args)).getDataItems();

                callbackResults = executeCallback(respFunc.getString(), callbackArgs);
            }

            co.nstant.in.cbor.model.Map req2 = new co.nstant.in.cbor.model.Map();
            req2.put(k_func, convertArg("_*executed*_"));
            co.nstant.in.cbor.model.Array v_args1 = new co.nstant.in.cbor.model.Array();
            for(int i = 0; i < callbackResults.size(); i++)
                v_args1.add(callbackResults.get(i));
            req2.put(k_args, v_args1);
            req2.put(k_uuid, convertArg(this.uuid));
            req2.put(k_lang, convertArg("java"));
            req2.put(k_ver, convertArg(this.VERSION));
            req2.put(k_argsl, convertArg(callbackResults.size()));
            if(this.verbose >= 1)
                System.out.println("Sending: " + req2.toString());
            this.send(req2);

            rep = (co.nstant.in.cbor.model.Map)this.recv();
            if(this.verbose >= 1)
                System.out.println("Received: " + rep.toString());
        }

        if(rep.getKeys().contains(k_err))
        {
            DataItem error = rep.get(k_err);
            throw new RuntimeException(error != null ? toString(error) : "Unknown error");
        }

        DataItem ret = rep.get(k_ret);
        // XXX: empty 'ret' comes in as a Map
        if(ret instanceof co.nstant.in.cbor.model.Map)
            return java.util.Collections.emptyList();
        return ((co.nstant.in.cbor.model.Array)ret).getDataItems();

    }

    public void registerCallback(String funcName, Function<Object[], Object[]> callback)
    {
        if(callbacks.containsKey(funcName))
            throw new RuntimeException("A callback is already registered for function: " + funcName);
        callbacks.put(funcName, callback);
    }

    private List<DataItem> executeCallback(String funcName, List<DataItem> args)
    {
        if(callbacks.containsKey(funcName))
        {
            Function<Object[], Object[]> callback = callbacks.get(funcName);
            return convertArgs(callback.apply(toObjects(args).toArray()));
        }
        else
            return new ArrayList<DataItem>(); // we cannot raise an error: e.g. a custom UI async callback cannot be assigned to a specific client
    }

    public DataItem convertArg(Boolean arg)
    {
        return convertArg(arg.booleanValue());
    }

    public DataItem convertArg(String arg)
    {
        return new co.nstant.in.cbor.model.UnicodeString(arg);
    }

    public DataItem convertArg(Integer arg)
    {
        return convertArg(arg.longValue());
    }

    public DataItem convertArg(Long arg)
    {
        return convertArg(arg.longValue());
    }

    public DataItem convertArg(Float arg)
    {
        return convertArg(arg.floatValue());
    }

    public DataItem convertArg(Double arg)
    {
        return convertArg(arg.doubleValue());
    }

    public DataItem convertArg(BigInteger arg)
    {
        if(arg.compareTo(BigInteger.ZERO) < 0)
            return new co.nstant.in.cbor.model.NegativeInteger(arg);
        else
            return new co.nstant.in.cbor.model.UnsignedInteger(arg);
    }

    public DataItem convertArg(Map arg)
    {
        throw new RuntimeException("unsupported conversion");
    }

    public DataItem convertArg(List arg)
    {
        co.nstant.in.cbor.model.Array array = new co.nstant.in.cbor.model.Array();
        for(Object arg1 : arg)
            array.add(convertArg(arg1));
        return array;
    }

    public DataItem convertArg(boolean arg)
    {
        if(arg) return co.nstant.in.cbor.model.SimpleValue.TRUE;
        else return co.nstant.in.cbor.model.SimpleValue.FALSE;
    }

    public DataItem convertArg(long arg)
    {
        if(arg < 0)
            return new co.nstant.in.cbor.model.NegativeInteger(arg);
        else
            return new co.nstant.in.cbor.model.UnsignedInteger(arg);
    }

    public DataItem convertArg(float arg)
    {
        return new co.nstant.in.cbor.model.SinglePrecisionFloat(arg);
    }

    public DataItem convertArg(double arg)
    {
        return new co.nstant.in.cbor.model.DoublePrecisionFloat(arg);
    }

    public DataItem convertArg(byte[] arg)
    {
        return new co.nstant.in.cbor.model.ByteString(arg);
    }

    public DataItem convertArg(Object arg)
    {
        if(arg instanceof Boolean)
            return convertArg((Boolean)arg);
        if(arg instanceof String)
            return convertArg((String)arg);
        if(arg instanceof Integer)
            return convertArg((Integer)arg);
        if(arg instanceof Long)
            return convertArg((Long)arg);
        if(arg instanceof Float)
            return convertArg((Float)arg);
        if(arg instanceof Double)
            return convertArg((Double)arg);
        if(arg instanceof BigInteger)
            return convertArg((BigInteger)arg);
        if(arg instanceof List)
            return convertArg((List)arg);
        if(arg instanceof Map)
            return convertArg((Map)arg);
        if(arg.getClass().isArray())
            return convertArg((byte[])arg);
        return null;
    }

    public void convertArgs(List<DataItem> outArgs, Object[] inArgs)
    {
        for(int i = 0; i < inArgs.length; i++)
            outArgs.add(convertArg(inArgs[i]));
    }

    public List<DataItem> convertArgs(Object[] inArgs)
    {
        List<DataItem> outArgs = new ArrayList<DataItem>();
        for(int i = 0; i < inArgs.length; i++)
            outArgs.add(convertArg(inArgs[i]));
        return outArgs;
    }

    public Boolean toBoolean(DataItem item)
    {
        co.nstant.in.cbor.model.SimpleValue v = (co.nstant.in.cbor.model.SimpleValue)item;
        co.nstant.in.cbor.model.SimpleValueType t = v.getSimpleValueType();
        if(t == co.nstant.in.cbor.model.SimpleValueType.TRUE) return Boolean.TRUE;
        if(t == co.nstant.in.cbor.model.SimpleValueType.FALSE) return Boolean.FALSE;
        return null;
    }

    public BigInteger toBigInteger(DataItem item)
    {
        return ((co.nstant.in.cbor.model.Number)item).getValue();
    }

    public int toInt(DataItem item)
    {
        return toBigInteger(item).intValue();
    }

    public long toLong(DataItem item)
    {
        return toBigInteger(item).longValue();
    }

    public double toFloat(DataItem item)
    {
        return ((co.nstant.in.cbor.model.AbstractFloat)item).getValue();
    }

    public double toDouble(DataItem item)
    {
        return ((co.nstant.in.cbor.model.DoublePrecisionFloat)item).getValue();
    }

    public String toString(DataItem item)
    {
        return ((co.nstant.in.cbor.model.UnicodeString)item).getString();
    }

    public byte[] toBytes(DataItem item)
    {
        return ((co.nstant.in.cbor.model.ByteString)item).getBytes();
    }

    public Map toMap(DataItem item)
    {
        co.nstant.in.cbor.model.Map m = (co.nstant.in.cbor.model.Map)item;
        Map ret = new HashMap();
        for(DataItem k : m.getKeys())
            ret.put(toObject(k), toObject(m.get(k)));
        return ret;
    }

    public Object toObject(DataItem item)
    {
        if(item instanceof co.nstant.in.cbor.model.SimpleValue) {
            Boolean b = toBoolean(item);
            if(b != null) return b;
        }
        if(item instanceof co.nstant.in.cbor.model.ByteString)
            return toBytes(item);
        if(item instanceof co.nstant.in.cbor.model.AbstractFloat)
            return toFloat(item);
        if(item instanceof co.nstant.in.cbor.model.DoublePrecisionFloat)
            return toDouble(item);
        if(item instanceof co.nstant.in.cbor.model.UnicodeString)
            return toString(item);
        if(item instanceof co.nstant.in.cbor.model.Number)
            return new Long(toLong(item));
        if(item instanceof co.nstant.in.cbor.model.Array)
            return toObjects(((co.nstant.in.cbor.model.Array)item).getDataItems());
        if(item instanceof co.nstant.in.cbor.model.Map)
            return toMap(item);
        return null;
    }

    public List<Object> toObjects(List<DataItem> items)
    {
        List<Object> ret = new ArrayList<Object>();
        for(DataItem item : items)
            ret.add(toObject(item));
        return ret;
    }

    public Object[] call(String func, Object[] args) throws CborException
    {
        return toObjects(callCbor(func, convertArgs(args))).toArray();
    }

    public void setStepping() throws CborException
    {
        setStepping(true);
    }

    public void setStepping(Boolean enable) throws CborException
    {
        call("sim.setStepping", new Object[]{enable});
    }

    public void step() throws IOException, CborException
    {
        step(true);
    }

    public void step(Boolean wait) throws IOException, CborException
    {
        call("sim.step", new Object[]{wait});
    }

    public RemoteAPIObjects getObject()
    {
        if(objs == null)
            objs = new RemoteAPIObjects(this);
        return objs;
    }
    
    private Map<String, Function<Object[], Object[]>> callbacks = new HashMap<>();
    private int verbose = -1;
    ZContext context;
    ZMQ.Socket rpcSocket;
    String uuid;
    int VERSION = 2;
    RemoteAPIObjects objs = null;
}
